### PR TITLE
Provide a common infra for maintenance jobs in perf testing

### PR DESCRIPTION
**Pull request by chizhg**
**What this PR does, why we need it**:
This PR is part of the work to provide a common infra for maintenance jobs for perf testing:
1. Add `performance-tests.sh`, each repo can source this script, and overwrite function `update_knative` and `update_benchmark` to define their own logic for benchmarking setup.
Each repo will need to set up 3 Prow jobs by using the script:
- `--update-clusters`: update the clusters with the latest Knative code, needs to be run as a periodic jobs with relatively short interval, e.g. one hour
- `--recreate-clusters`: delete and recreate clusters for the benchmarks, in case there is GKE update or broken node for the cluster, needs to be run as a periodic job with relatively long interval, e.g. one day
- `--recreate-clusters`: reconcile the clusters to match with the cluster configs for the benchmarks if there is any config change, needs to be run as a postsubmit job
2. Create a Go library to perform all cluster operations, by using the `gke` package in `pkg/test`. Similar as https://github.com/knative/test-infra/pull/1460, the library will be ported to `pkg`

### CODE REVIEW

File: perf-cluster-operation/main.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (adrcunha) [inline perf-cluster-operation/main.go]: This is a tool, it should be under /tools. Also, a shorter name would be nice. What about "perf-tool", "perf-ops", or something like that?

File: scripts/performance-tests.sh [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    # Update all resources installed on the cluster
+    update_cluster ${name} ${zone}
+  done
+  header "Done updating all clusters"
+}
+
+# Delete the old clusters related to the current repo, and recreate them with the same configuration.
+function recreate_clusters() {
+  header "Recreating clusters for ${REPO_NAME}"
+  go run $GOPATH/src/knative.dev/test-infra/pkg/perf-cluster-operation \
```
Comment (adrcunha) [inline scripts/performance-tests.sh]: This is an absolute location, meaning that any repo sourcing this script must have test-infra checked out in order to run this command. I believe you intend to have the tool vendored instead. Here and everywhere.

**Comment by adrcunha** [comment]
Why not adding the library (step 2 in the initial comment) to knative/pkg directly, instead of having it here first, then migrating?

File: perf-cluster-operation/pkg/cluster.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+			}
+		}()
+		// remove the cluster from clusterConfigs as it's already been handled
+		delete(clusterConfigs, cluster.Name)
+	}
+
+	// handle all other cluster configs
+	for name, config := range clusterConfigs {
+		wg.Add(1)
+		newName, newConfig := name, config
```
Comment (chaodaiG) [inline perf-cluster-operation/pkg/cluster.go]:
```suggestion
		name, config := name, config
```

File: perf-cluster-operation/pkg/cluster.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	var err error
+	for i := 0; i < retryTimes; i++ {
+		if err = gc.ops.DeleteCluster(gcpProject, region, zone, cluster.Name); err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return fmt.Errorf(
+			"failed deleting cluster %q in %q after retrying %d times: %v",
+			cluster.Name, gke.GetClusterLocation(region, zone), retryTimes, err)
```
Comment (chaodaiG) [inline perf-cluster-operation/pkg/cluster.go]:
```suggestion
			cluster.Name, cluster.Location, retryTimes, err)
```

File: perf-cluster-operation/pkg/cluster.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		NodeType:    config.NodeType,
+		Addons:      strings.Split(config.Addons, ","),
+	}
+	creq, err := gke.NewCreateClusterRequest(req)
+	if err != nil {
+		return fmt.Errorf("cannot create cluster with request %v: %v", req, err)
+	}
+
+	region, zone := gke.RegionZoneFromLoc(config.Location)
+	for i := 0; i < retryTimes; i++ {
```
Comment (chaodaiG) [inline perf-cluster-operation/pkg/cluster.go]: if creation failed after it started, then retry for sure will fail since the old cluster still exists

**Comment by chizhg** [comment]
> Why not adding the library (step 2 in the initial comment) to knative/pkg directly, instead of having it here first, then migrating?

I got this idea from Chao, first we put all the code here and get it reviewed, and manually test it to make sure it works end-to-end, then we create a PR in `pkg` and merge all Go code.

In this way we can guarantee the code in `pkg` and `test-infra` are compatible,  and do not need to make further changes in `pkg`.

File: perf-cluster-operation/pkg/benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		)
+	}
+}
+
+func TestBenchmarkNameForCluster(t *testing.T) {
+	testCases := []struct {
+		clusterName           string
+		repo                  string
+		expectedBenchmarkName string
+	}{{
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: That's a very good point..
The reason that I'm adding `repo` as the cluster name prefix is, if two repos are using a same gcloud project, we can distinguish which repo owns which clusters. But with this approach, there is no way to differentiate for `eventing` and `eventing-contrib`. 
So we'll have to use two different projects for the two repos. However, I think it's okay since it's what we are doing now for `serving` and `eventing`, and creating a new gcloud project is not a big deal.

File: perf-cluster-operation/pkg/benchmark_test.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		expectedError:      true,
+	}}
+
+	for _, tc := range testCases {
+		benchmarks, err := benchmarkNames(tc.benchmarkRoot)
+		if !reflect.DeepEqual(tc.expectedBenchmarks, benchmarks) {
+			t.Fatalf("expected to get benchmarks %v under %q, but got %v",
+				tc.expectedBenchmarks, tc.benchmarkRoot, benchmarks)
+		}
+		if tc.expectedError && err == nil {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: I personally feel it too strict for a unit test to also test the error message, since in most scenarios the error message is not that important.

File: perf-cluster-operation/pkg/benchmark.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+func benchmarkNameForCluster(clusterName, repo string) string {
+	if !clusterBelongsToRepo(clusterName, repo) {
+		return ""
+	}
+	return strings.TrimPrefix(clusterName, repo+"-")
+}
+
+// clusterBelongsToRepo determines if the cluster belongs to the repo, by checking if it has the repo prefix.
+func clusterBelongsToRepo(clusterName, repo string) bool {
+	return strings.HasPrefix(clusterName, repo+"-")
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: As I mentioned in the following comment, it is indeed a problem for `eventing` and `eventing-contrib`, but there is a workaround by using two different projects for them.

If we really want to differentiate them, we'll need a hack way, since cluster name only allows `lowercase letters, numbers, or hyphens`. WDYT?

File: perf-cluster-operation/pkg/benchmark.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	return clusters, nil
+}
+
+// clusterConfigForBenchmark returns the cluster config for the given benchmark.
+//
+// Under each benchmark folder, we can put a cluster.yaml file that follows the scheme we define
+// in ClusterConfig struct, in which we specify configuration of the cluster that we use to run the benchmark.
+// If there is no such config file, or the config file is malformed, default config will be used.
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Yes, I actually have a unit test for it, see the file under `test-benchmark2`

File: perf-cluster-operation/pkg/benchmark.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	// clusters is a map of cluster configs
+	// key is the cluster name, value is the cluster config
+	clusters := make(map[string]ClusterConfig)
+	benchmarkNames, err := benchmarkNames(benchmarkRoot)
+	if err != nil {
+		return clusters, err
+	}
+
+	for _, benchmarkName := range benchmarkNames {
+		clusterConfig := clusterConfigForBenchmark(benchmarkRoot, benchmarkName)
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Hmm, just curious, is it one of Go best practice to use short var names? I think long vars are more readable.

File: perf-cluster-operation/pkg/benchmark.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	"strings"
+
+	yaml "gopkg.in/yaml.v2"
+	"knative.dev/test-infra/shared/common"
+)
+
+const (
+	// clusterConfigFile is the config file needs to be put under the benchmark folder if we want to config
+	// the cluster that runs the benchmark, it must follow the scheme defined as GKECluster here.
+	clusterConfigFile = "cluster.yaml"
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Do you mean the file name or the yaml scheme as well?

File: perf-cluster-operation/pkg/benchmark.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (adrcunha) [reply perf-cluster-operation/pkg/benchmark.go]: https://github.com/golang/go/wiki/CodeReviewComments#variable-names

```
Variable names in Go should be short rather than long. This is especially true for local variables with limited scope. Prefer c to lineCount. Prefer i to sliceIndex.
```

File: perf-cluster-operation/pkg/benchmark.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+// TODO(chizhg): it's currently not used, use it in the cluster creation retry logic.
+var backupLocations = []string{"us-west1", "us-west2", "us-east1"}
+
+// GKECluster saves the config information for the GKE cluster
+type GKECluster struct {
+	Config ClusterConfig `yaml:"GKECluster,omitempty"`
+}
+
+// ClusterConfig is config for the cluster
+type ClusterConfig struct {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Sure, we can. But we'll need some changes to the `gke.Request` first, e.g. renaming it to `gke.ClusterConfig`, adding yaml scheme, etc.

I would prefer to keep it as it is for now, but do the necessary changes after we move all the Go code to `pkg`.

File: perf-cluster-operation/pkg/benchmark.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply perf-cluster-operation/pkg/benchmark.go]: I mean the location of this file, can consumer of this package provide an absolute path of the yaml?

File: perf-cluster-operation/pkg/benchmark.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply perf-cluster-operation/pkg/benchmark.go]: The unit test is certainly a bad config. What if you accidentally committed something that has a slight indentation error, do you still want it fall back to use default?

File: perf-cluster-operation/main.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+		if err := client.RecreateClusters(gcpProjectName, repoName, benchmarkRootFolder); err != nil {
+			log.Fatalf("Failed recreating clusters for repo %q: %v", repoName, err)
+		} else {
+			log.Printf("Done with recreating clusters for repo %q", repoName)
+		}
+	case "reconcile":
+		if err := client.ReconcileClusters(gcpProjectName, repoName, benchmarkRootFolder); err != nil {
+			log.Fatalf("Failed reconciling clusters for repo %q: %v", repoName, err)
+		} else {
+			log.Printf("Done with reconciling clusters for repo %q", repoName)
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/main.go]: There will be a grammar mistake if I make this change - "Done with reconcile" :(

File: perf-cluster-operation/pkg/benchmark.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply perf-cluster-operation/pkg/benchmark.go]: can we check it from the end? Hopefully there is no `-` in your benchmark names

File: perf-cluster-operation/main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply perf-cluster-operation/main.go]: lol, just leave them as they are then

File: perf-cluster-operation/main.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+	client, err := pkg.NewClient()
+	if err != nil {
+		log.Fatalf("failed setting up GKE client, cannot proceed: %v", err)
+	}
+	operation := args[0]
+	switch operation {
+	case "recreate":
+		if err := client.RecreateClusters(gcpProjectName, repoName, benchmarkRootFolder); err != nil {
+			log.Fatalf("Failed recreating clusters for repo %q: %v", repoName, err)
+		} else {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/main.go]: Removed.

File: perf-cluster-operation/main.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+			log.Printf("Done with recreating clusters for repo %q", repoName)
+		}
+	case "reconcile":
+		if err := client.ReconcileClusters(gcpProjectName, repoName, benchmarkRootFolder); err != nil {
+			log.Fatalf("Failed reconciling clusters for repo %q: %v", repoName, err)
+		} else {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/main.go]: Removed.

File: perf-cluster-operation/pkg/benchmark.go [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+			Addons:    defaultAddons,
+		},
+	}
+
+	configFile := filepath.Join(benchmarkRoot, benchmarkName, clusterConfigFile)
+	if common.FileExists(configFile) {
+		contents, err := ioutil.ReadFile(configFile)
+		if err == nil {
+			if err := yaml.Unmarshal(contents, &gkeCluster); err != nil {
+				log.Printf("Failed to parse the config file %q, default config will be used", configFile)
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Done.

File: perf-cluster-operation/pkg/benchmark.go [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+			}
+		} else {
+			log.Printf("Failed to read the config file %q, default config will be used", configFile)
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Done. Thanks.

File: perf-cluster-operation/pkg/benchmark_test.go [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+		expectedError:      false,
+	}, {
+		benchmarkRoot:      "non-existed-dir",
+		expectedBenchmarks: []string{},
+		expectedError:      true,
+	}}
+
+	for _, tc := range testCases {
+		benchmarks, err := benchmarkNames(tc.benchmarkRoot)
+		if !reflect.DeepEqual(tc.expectedBenchmarks, benchmarks) {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: Thanks for the pointer. Replaced it.

File: perf-cluster-operation/pkg/benchmark_test.go [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+	}, {
+		benchmarkRoot: "testdir",
+		benchmarkName: "test-benchmark4",
+		expectedClusterConfig: ClusterConfig{
+			Location: defaultLocation, NodeCount: defaultNodeCount, NodeType: defaultNodeType, Addons: defaultAddons},
+	}}
+
+	for _, tc := range testCases {
+		clusterConfig := clusterConfigForBenchmark(tc.benchmarkRoot, tc.benchmarkName)
+		if !reflect.DeepEqual(tc.expectedClusterConfig, clusterConfig) {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: Done.

File: perf-cluster-operation/pkg/benchmark_test.go [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+			t.Fatalf("expected to get cluster config %v for benchmark %q under %q, but got %v",
+				tc.expectedClusterConfig, tc.benchmarkName, tc.benchmarkRoot, clusterConfig)
+		}
+	}
+}
+
+func TestClusterNameForBenchmark(t *testing.T) {
+	repo, benchmarkName, expectedName := "serving", "load-test", "serving-load-test"
+	gotName := clusterNameForBenchmark(benchmarkName, repo)
+	if gotName != expectedName {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: Done.

File: perf-cluster-operation/pkg/benchmark_test.go [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+		expectedBenchmarkName: "",
+	}, {
+		clusterName:           "serving--load-test",
+		repo:                  "serving",
+		expectedBenchmarkName: "-load-test",
+	}}
+
+	for _, tc := range testCases {
+		benchmarkName := benchmarkNameForCluster(tc.clusterName, tc.repo)
+		if benchmarkName != tc.expectedBenchmarkName {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: Done.

File: perf-cluster-operation/pkg/cluster.go (hunk 5 above)
Comment (chizhg) [reply perf-cluster-operation/pkg/cluster.go]: Yes that's true.. So we probably should delete the old cluster and then retry, based on the error type? Does this problem also apply to cluster setup for e2e tests?

File: perf-cluster-operation/main.go [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+		log.Fatal("only one operation name can be provided as the arg")
+	}
+
+	client, err := pkg.NewClient()
+	if err != nil {
+		log.Fatalf("failed setting up GKE client, cannot proceed: %v", err)
+	}
+	operation := args[0]
+	switch operation {
+	case "recreate":
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/main.go]: Changed to const.

File: perf-cluster-operation/pkg/benchmark.go [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+	}
+	return names, nil
+}
+
+// benchmarkClusters returns the cluster configs for all benchmarks.
+func benchmarkClusters(repo, benchmarkRoot string) (map[string]ClusterConfig, error) {
+	// clusters is a map of cluster configs
+	// key is the cluster name, value is the cluster config
+	clusters := make(map[string]ClusterConfig)
+	benchmarkNames, err := benchmarkNames(benchmarkRoot)
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: After this line, `benchmarkNames` becomes `[]string` type, and you will not be able to call the `benchmarkNames` within this function.

File: perf-cluster-operation/pkg/cluster.go [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+
+	container "google.golang.org/api/container/v1beta1"
+)
+
+const (
+	// the maximum retry times if there is an error in cluster operation
+	retryTimes = 3
+
+	// known cluster status
+	statusProvisioning = "PROVISIONING"
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/cluster.go]: Sure, we can move it to `gke` package after all code is ported to `pkg`.

File: perf-cluster-operation/pkg/cluster.go [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+	statusRunning      = "RUNNING"
+	statusStopping     = "STOPPING"
+)
+
+type gkeClient struct {
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/cluster.go]: Yes.. I'll add the tests in my next PR.

File: perf-cluster-operation/main.go [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+	args := flag.Args()
+	if len(args) != 1 {
+		log.Fatal("only one operation name can be provided as the arg")
+	}
+
+	client, err := pkg.NewClient()
+	if err != nil {
+		log.Fatalf("failed setting up GKE client, cannot proceed: %v", err)
+	}
+	operation := args[0]
```
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/main.go]: This is the best solution I can find without using `cobra`, if you have better ideas we can discuss

File: perf-cluster-operation/pkg/benchmark.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: There will be `-` in the benchmark names. But adding as the postfix can indeed solve the problem, since there can be `eventing-contrib`, but I believe there will be less opportunity to have `contrib-eventing`...
I can make the change if you are also fine with postfix.

File: perf-cluster-operation/pkg/benchmark.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply perf-cluster-operation/pkg/benchmark.go]: sgtm

File: perf-cluster-operation/pkg/benchmark_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (chaodaiG) [reply perf-cluster-operation/pkg/benchmark_test.go]: sgtm

File: perf-cluster-operation/pkg/benchmark.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: I think it would be better to have a common standard for organizing the benchmark folder, and add it to the instruction of how to write a new benchmark. So I would prefer to only define it here.

File: perf-cluster-operation/pkg/benchmark.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Not sure how double `-` can solve this problem, let's sync up offline.

File: perf-cluster-operation/pkg/cluster.go (hunk 5 above)
Comment (chizhg) [reply perf-cluster-operation/pkg/cluster.go]: Added the deletion logic.
I think `Wait` won't finish when the cluster is still in `PROVISIONING` state, unless something really bad happens, then some operations will fail anyway.

**Comment by chaodaiG** [comment]
yeah, it's not designed to be thread-safe, you'll need to make some changes
to it

File: perf-cluster-operation/pkg/benchmark.go [hunk 27]
```diff
@@ -124,7 +124,7 @@ func clusterConfigForBenchmark(benchmarkName, benchmarkRoot string) ClusterConfi
 
 // clusterNameForBenchmark prepends repo name to the benchmark name, and use it as the cluster name.
 func clusterNameForBenchmark(benchmarkName, repo string) string {
-	return fmt.Sprintf("%s-%s", repo, benchmarkName)
+	return fmt.Sprintf("%s--%s", repo, benchmarkName)
```
Comment (chaodaiG) [inline perf-cluster-operation/pkg/benchmark.go]: I think this deserves a `repoPrefix(repo string) string` function

File: vendor/knative.dev/pkg/test/gke/fake/client.go [hunk 28]
```diff
@@ -45,6 +46,8 @@ type GKESDKClient struct {
 	opNumber int
 	// A lookup table for determining ops statuses
 	OpStatus map[string]string
+
+	mutex sync.Mutex
```
Comment (chaodaiG) [inline vendor/knative.dev/pkg/test/gke/fake/client.go]: this is not part of this PR

Comment (chaodaiG) [reply vendor/knative.dev/pkg/test/gke/fake/client.go]: you may need to merge or rebase

File: perf-cluster-operation/pkg/benchmark_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark_test.go]: Done

File: perf-cluster-operation/pkg/benchmark.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Done with changing to double `-`

File: perf-cluster-operation/main.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (chizhg) [reply perf-cluster-operation/main.go]: Done.

File: perf-cluster-operation/pkg/benchmark.go (hunk 27 above)
Comment (chizhg) [reply perf-cluster-operation/pkg/benchmark.go]: Done.

**Comment by chaodaiG** [comment]
/lgtm
/approve
/hold
please double check to make sure it's not flaky before merging

File: perf-cluster-operation/main.go (hunk 1 above)
Comment (chizhg) [reply perf-cluster-operation/main.go]: Moved to `tools/perf-ops`. Thanks.

File: scripts/performance-tests.sh (hunk 2 above)
Comment (chizhg) [reply scripts/performance-tests.sh]: Changed to `go run knative.dev/test-infra/tools/perf-ops`

**Comment by chizhg** [comment]
> /lgtm
> /approve
> /hold
> please double check to make sure it's not flaky before merging

I've run it with `-count 100` again and all tests passed.

**Comment by chizhg** [comment]
Close this PR in favor of https://github.com/knative/test-infra/pull/1478 and https://github.com/knative/pkg/pull/776 in `knative/pkg`

/close
